### PR TITLE
Support various datetime types as input

### LINKED ISSUE #242: Support datetime data types as input
**Description of the problem**
I want to plot some datetime data on a map. However, gmt-python doesn't accept string as input.

**Full code that generated the error**

```python
import pygmt

x = ['2008-01-01', '2012-01-01']
y = [5, 5]

fig = pygmt.Figure()
fig.plot(x, y, J='X10c/5c', R='2005-01-01/2015-01-01/0/10', B=True, W='2p')
fig.show(method='external')
```

**Full error message**

```
Traceback (most recent call last):
  File "test.py", line 7, in <module>
    fig.plot(x, y, J='X10c/5c', R='2005-01-01/2015/01-01/0/10', B=True, W='2p')
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/helpers/decorators.py", line 199, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/helpers/decorators.py", line 294, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/base_plotting.py", line 344, in plot
    with file_context as fname:
  File "/Users/seisman/.anaconda/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/clib/session.py", line 1067, in virtualfile_from_vectors
    self.put_vector(dataset, column=col, vector=array)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/clib/session.py", line 754, in put_vector
    gmt_type = self._check_dtype_and_dim(vector, ndim=1)
  File "/Users/seisman/Gits/gmt/gmt-python/gmt/clib/session.py", line 703, in _check_dtype_and_dim
    "Unsupported numpy data type '{}'.".format(array.dtype.name)
gmt.exceptions.GMTInvalidInput: Unsupported numpy data type 'str320'.
```

**System information**

* Operating system: macOS 10.14
* Python installation (Anaconda, system, ETS): Anaconda
* Version of GMT: 6.0.0_15a9160
* Version of Python: 3.7.0
* Version of this package: latest
* If using conda, paste the output of `conda list` below:

<details>
<summary>output of conda list</summary>
<pre>
# packages in environment at /Users/seisman/.anaconda:
#
# Name                    Version                   Build  Channel
_ipyw_jlab_nb_ext_conf    0.1.0                    py37_0
alabaster                 0.7.11                   py37_0
anaconda                  5.3.0                    py37_0
anaconda-client           1.7.2                    py37_0
anaconda-navigator        1.9.2                    py37_0
anaconda-project          0.8.2                    py37_0
appdirs                   1.4.3            py37h28b3542_0
appnope                   0.1.0                    py37_0
appscript                 1.0.1            py37h1de35cc_1
argh                      0.26.2                    <pip>
asn1crypto                0.24.0                   py37_0
astroid                   2.0.4                    py37_0
astropy                   3.0.4            py37h1de35cc_0
atomicwrites              1.2.1                    py37_0
attrs                     18.2.0           py37h28b3542_0
automat                   0.7.0                    py37_0
babel                     2.6.0                    py37_0
backcall                  0.1.0                    py37_0
backports                 1.0                      py37_1
backports.shutil_get_terminal_size 1.0.0                    py37_2
basemap                   1.2.0            py37h0acbc05_0
beautifulsoup4            4.6.3                    py37_0
bitarray                  0.8.3            py37h1de35cc_0
bkcharts                  0.2                      py37_0
black                     18.9b0                    <pip>
blas                      1.0                         mkl
blaze                     0.11.3                   py37_0
bleach                    2.1.4                    py37_0
blosc                     1.14.4               hd9629dc_0
bokeh                     0.13.0                   py37_0
boto                      2.49.0                   py37_0
bottleneck                1.2.1            py37h1d22016_1
bzip2                     1.0.6                h1de35cc_5
ca-certificates           2018.03.07                    0
cartopy                   0.16.0           py37h9263bd1_0
certifi                   2018.8.24                py37_1
cffi                      1.11.5           py37h6174b99_1
cftime                    1.0.0b1          py37h1d22016_0
chardet                   3.0.4                    py37_1
click                     6.7                      py37_0
cloud-sptheme             1.9.4                     <pip>
cloudpickle               0.5.5                    py37_0
clyent                    1.2.2                    py37_1
cmarkgfm                  0.4.2                     <pip>
colorama                  0.3.9                    py37_0
conda                     4.5.11                py37_1000    conda-forge
conda-build               3.15.1                   py37_0
conda-env                 2.6.0                         1
constantly                15.1.0           py37h28b3542_0
contextlib2               0.5.5                    py37_0
cryptography              2.3.1            py37hdbc3d79_0
curl                      7.61.0               ha441bb4_0
cycler                    0.10.0                   py37_0
cython                    0.28.5           py37h0a44026_0
cytoolz                   0.9.0.1          py37h1de35cc_1
dask                      0.19.1                   py37_0
dask-core                 0.19.1                   py37_0
datashape                 0.5.4                    py37_1
dbus                      1.13.2               h760590f_1
decorator                 4.3.0                    py37_0
defusedxml                0.5.0                    py37_1
distributed               1.23.1                   py37_0
docutils                  0.14                     py37_0
entrypoints               0.2.3                    py37_2
et_xmlfile                1.0.1                    py37_0
expat                     2.2.6                h0a44026_0
fastcache                 1.0.2            py37h1de35cc_2
filelock                  3.0.8                    py37_0
flask                     1.0.2                    py37_1
flask-cors                3.0.6                    py37_0
freetype                  2.9.1                hb4e5f40_0
future                    0.16.0                    <pip>
future                    0.17.0                py37_1000    conda-forge
geographiclib             1.49                      <pip>
geos                      3.6.2                h5470d99_2
get_terminal_size         1.0.0                h7520d66_0
gettext                   0.19.8.1             h15daf44_3
gevent                    1.3.6            py37h1de35cc_0
glib                      2.56.2               hd9629dc_0
glob2                     0.6                      py37_0
gmp                       6.1.2                hb37e062_1
gmpy2                     2.0.8            py37h6ef4df4_2
gmt-python                0.1a3+126.g4617492.dirty           <pip>
greenlet                  0.4.15           py37h1de35cc_0
guzzle-sphinx-theme       0.7.11                    <pip>
h5py                      2.8.0            py37h878fce3_3
hdf4                      4.2.13               h39711bb_2
hdf5                      1.10.2               hfa1e0ec_1
heapdict                  1.0.0                    py37_2
html5lib                  1.0.1                    py37_0
hyperlink                 18.0.0                   py37_0
icu                       58.2                 h4b95b61_1
idna                      2.7                      py37_0
imageio                   2.4.1                    py37_0
imagesize                 1.1.0                    py37_0
incremental               17.5.0                   py37_0
intel-openmp              2019.0                      118
ipykernel                 4.9.0                    py37_1
ipython                   6.5.0                    py37_0
ipython_genutils          0.2.0                    py37_0
ipywidgets                7.4.1                    py37_0
isort                     4.3.4                    py37_0
itsdangerous              0.24                     py37_1
jbig                      2.1                  h4d881f8_0
jdcal                     1.4                      py37_0
jedi                      0.12.1                   py37_0
jinja2                    2.10                     py37_0
jpeg                      9b                   he5867d9_2
jsonschema                2.6.0                    py37_0
jupyter                   1.0.0                    py37_7
jupyter_client            5.2.3                    py37_0
jupyter_console           5.2.0                    py37_1
jupyter_core              4.4.0                    py37_0
jupyterlab                0.34.9                   py37_0
jupyterlab_launcher       0.13.1                   py37_0
keyring                   13.2.1                   py37_0
kiwisolver                1.0.1            py37h0a44026_0
lazy-object-proxy         1.3.1            py37h1de35cc_2
libcurl                   7.61.0               hf30b1f0_0
libcxx                    4.0.1                h579ed51_0
libcxxabi                 4.0.1                hebd6815_0
libedit                   3.1.20170329         hb402a30_2
libffi                    3.2.1                h475c297_4
libgfortran               3.0.1                h93005f0_2
libiconv                  1.15                 hdd342a3_7
libnetcdf                 4.6.1                h4e6abe9_1
libpng                    1.6.34               he12f830_0
libsodium                 1.0.16               h3efe00b_0
libssh2                   1.8.0                h322a93b_4
libtiff                   4.0.9                hcb84e12_2
libxml2                   2.9.8                hab757c2_1
libxslt                   1.1.32               hb819dd2_0
livereload                2.5.2                     <pip>
llvmlite                  0.24.0           py37hc454e04_0
locket                    0.2.0                    py37_1
lxml                      4.2.5            py37hef8c89e_0
lzo                       2.10                 h362108e_2
markupsafe                1.0              py37h1de35cc_1
matplotlib                2.2.3            py37h54f8f79_0
mccabe                    0.6.1                    py37_1
mistune                   0.8.3            py37h1de35cc_1
mkl                       2019.0                      118
mkl-service               1.1.2            py37h6b9c3cc_5
mkl_fft                   1.0.4            py37h5d10147_1
mkl_random                1.0.1            py37h5d10147_1
more-itertools            4.3.0                    py37_0
mpc                       1.1.0                h6ef4df4_1
mpfr                      4.0.1                h3018a27_3
mpmath                    1.0.0                    py37_2
msgpack-python            0.5.6            py37h04f5b5a_1
multipledispatch          0.6.0                    py37_0
navigator-updater         0.2.1                    py37_0
nbconvert                 5.4.0                    py37_1
nbformat                  4.4.0                    py37_0
ncurses                   6.1                  h0a44026_0
netcdf4                   1.4.1            py37h08833f9_0
networkx                  2.1                      py37_0
nltk                      3.3.0                    py37_0
nose                      1.3.7                    py37_2
notebook                  5.6.0                    py37_0
numba                     0.39.0           py37h6440ff4_0
numexpr                   2.6.8            py37h1dc9127_0
numpy                     1.15.1           py37h6a91979_0
numpy-base                1.15.1           py37h8a80b8c_0
numpydoc                  0.8.0                    py37_0
obspy                     1.1.0            py37h39e3cac_2    conda-forge
obspy                     1.1.0                     <pip>
odo                       0.5.1                    py37_0
olefile                   0.46                     py37_0
openpyxl                  2.5.6                    py37_0
openssl                   1.0.2p               h1de35cc_0
owslib                    0.17.0                   py37_0
packaging                 17.1                     py37_0
pandas                    0.23.4           py37h6440ff4_0
pandoc                    1.19.2.1             ha5e8f32_1
pandocfilters             1.4.2                    py37_1
parso                     0.3.1                    py37_0
partd                     0.3.8                    py37_0
path.py                   11.1.0                   py37_0
pathlib2                  2.3.2                    py37_0
pathtools                 0.1.2                     <pip>
patsy                     0.5.0                    py37_0
pcre                      8.42                 h378b8a2_0
pep8                      1.7.1                    py37_0
pexpect                   4.6.0                    py37_0
pickleshare               0.7.4                    py37_0
pillow                    5.2.0            py37hb68e598_0
pip                       10.0.1                   py37_0
pkginfo                   1.4.2                    py37_1
pluggy                    0.7.1            py37h28b3542_0
ply                       3.11                     py37_0
port-for                  0.3.1                     <pip>
proj4                     5.0.1                h1de35cc_0
prometheus_client         0.3.1            py37h28b3542_0
prompt_toolkit            1.0.15                   py37_0
psutil                    5.4.7            py37h1de35cc_0
ptyprocess                0.6.0                    py37_0
py                        1.6.0                    py37_0
pyasn1                    0.4.4            py37h28b3542_0
pyasn1-modules            0.2.2                    py37_0
pycodestyle               2.4.0                    py37_0
pycosat                   0.6.3            py37h1de35cc_0
pycparser                 2.18                     py37_1
pycrypto                  2.6.1            py37h1de35cc_9
pycurl                    7.43.0.2         py37hdbc3d79_0
pyepsg                    0.3.2                    py37_0
pyflakes                  2.0.0                    py37_0
pygments                  2.2.0                    py37_0
pylint                    2.1.1                    py37_0
pyodbc                    4.0.24           py37h0a44026_0
pyopenssl                 18.0.0                   py37_0
pyparsing                 2.2.0                    py37_1
pyproj                    1.9.5.1          py37h833a5d7_1
pyqt                      5.9.2            py37h655552a_2
pyshp                     1.2.12                   py37_0
pysocks                   1.6.8                    py37_0
pytables                  3.4.4            py37h13cba08_0
pytest                    3.8.0                    py37_0
pytest-arraydiff          0.2              py37h39e3cac_0
pytest-astropy            0.4.0                    py37_0
pytest-doctestplus        0.1.3                    py37_0
pytest-mpl                0.10                      <pip>
pytest-openfiles          0.3.0                    py37_0
pytest-remotedata         0.3.0                    py37_0
python                    3.7.0                hc167b69_0
python-dateutil           2.7.3                    py37_0
python.app                2                        py37_8
pytz                      2018.5                   py37_0
pywavelets                1.0.0            py37h1d22016_0
pyyaml                    3.13             py37h1de35cc_0
pyzmq                     17.1.2           py37h1de35cc_0
qt                        5.9.6                h45cd832_2
qtawesome                 0.4.4                    py37_0
qtconsole                 4.4.1                    py37_0
qtpy                      1.5.0                    py37_0
readline                  7.0                  h1de35cc_5
readme-renderer           22.0                      <pip>
requests                  2.19.1                   py37_0
requests-toolbelt         0.8.0                     <pip>
rope                      0.11.0                   py37_0
ruamel_yaml               0.15.46          py37h1de35cc_0
scikit-image              0.14.0           py37h0a44026_1
scikit-learn              0.19.2           py37h4f467ca_0
scipy                     1.1.0            py37h28f7352_1
seaborn                   0.9.0                    py37_0
send2trash                1.5.0                    py37_0
service_identity          17.0.0           py37h28b3542_0
setuptools                40.2.0                   py37_0
shapely                   1.6.4            py37h20de77a_0
simplegeneric             0.8.1                    py37_2
singledispatch            3.4.0.3                  py37_0
sip                       4.19.8           py37h0a44026_0
six                       1.11.0                   py37_1
snappy                    1.1.7                he62c110_3
snowballstemmer           1.2.1                    py37_0
sortedcollections         1.0.1                    py37_0
sortedcontainers          2.0.5                    py37_0
sphinx                    1.7.9                    py37_0
Sphinx                    1.8.1                     <pip>
sphinx-autobuild          0.7.1                     <pip>
sphinx-cjkspace           0.1.2                     <pip>
sphinx-intl               0.9.11                    <pip>
sphinx-rtd-theme          0.4.1                     <pip>
sphinxcontrib             1.0                      py37_1
sphinxcontrib-websupport  1.1.0                    py37_1
spyder                    3.3.1                    py37_1
spyder-kernels            0.2.6                    py37_0
sqlalchemy                1.2.11           py37h1de35cc_0
sqlite                    3.24.0               ha441bb4_0
statsmodels               0.9.0            py37h1d22016_0
sympy                     1.2                      py37_0
tblib                     1.3.2                    py37_0
terminado                 0.8.1                    py37_1
testpath                  0.3.1                    py37_0
tk                        8.6.8                ha441bb4_0
toml                      0.10.0                    <pip>
toolz                     0.9.0                    py37_0
tornado                   5.1              py37h1de35cc_0
tqdm                      4.26.0           py37h28b3542_0
traitlets                 4.3.2                    py37_0
twine                     1.12.1                    <pip>
twisted                   18.7.0           py37h1de35cc_1
unicodecsv                0.14.1                   py37_0
unixodbc                  2.3.7                h1de35cc_0
urllib3                   1.23                     py37_0
watchdog                  0.9.0                     <pip>
wcwidth                   0.1.7                    py37_0
webencodings              0.5.1                    py37_1
werkzeug                  0.14.1                   py37_0
wheel                     0.31.1                   py37_0
widgetsnbextension        3.4.1                    py37_0
wrapt                     1.10.11          py37h1de35cc_2
xarray                    0.10.8                   py37_0
xlrd                      1.1.0                    py37_1
xlsxwriter                1.1.0                    py37_0
xlwings                   0.11.8                   py37_0
xlwt                      1.3.0                    py37_0
xz                        5.2.4                h1de35cc_4
yaml                      0.1.7                hc338f04_2
zeromq                    4.2.5                h0a44026_1
zhon                      1.1.5                     <pip>
zict                      0.1.3                    py37_0
zlib                      1.2.11               hf3cbc9b_2
zope                      1.0                      py37_1
zope.interface            4.5.0            py37h1de35cc_0
</pre>
</details>

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This PR allows PyGMT accept vectors of various datetime types so that PyGMT can plot datetime axis. 

Vectors of following datetime types are supported:

- [x] [numpy.datetime64](https://numpy.org/doc/stable/reference/arrays.datetime.html)
- [x] [pandas.DateTimeIndex](https://pandas.pydata.org/docs/user_guide/timeseries.html)
- [x] Raw datetime strings in both ISO and non-ISO formats, e.g., `2010-01-01`, `1/1/2018`, `Jul 5, 2019`
- [x] Python's [built-in datetime and date](https://docs.python.org/3/library/datetime.html)

TODO:

- [x] Wait for the GMT 6.1.0 release
- [ ] Add more tests
- [ ] Add a tutorial for calendar axis
- [x] Test for xarray.DataArray

Known issues but won't be fixed in this PR:

- [ ] The `region` argument doesn't work with numpy `datetime64` objects, have to convert to a string using [`np.datetime_as_string`](https://numpy.org/devdocs/reference/generated/numpy.datetime_as_string.html) in order to set the map frame bounds.
- [ ] Support [`timedelta64`](https://pandas.pydata.org/pandas-docs/stable/user_guide/timedeltas.html), e.g. GPS time from a certain epoch. 

Fixes #242.

**Example script:**
```python
import datetime
import pygmt
import numpy as np
import pandas as pd

fig = pygmt.Figure()
fig.basemap(projection="X15c/5c", region="2010-01-01/2020-01-01/0/10", frame=True)

# numpy.datetime64 types
x = np.array(
    ["2010-06-01", "2011-06-01T12", "2012-01-01T12:34:56"], dtype="datetime64"
)
y = [1.0, 2.0, 3.0]
fig.plot(x, y, style="c0.2c", pen="1p")

# pandas.DatetimeIndex
x = pd.date_range("2013", freq="YS", periods=3)
y = [4, 5, 6]
fig.plot(x, y, style="t0.2c", pen="1p")

# xarray.DataArray
x = xr.DataArray(data=pd.date_range(start="2015-03", freq="QS", periods=3))
y = [7.5, 6, 4.5]
fig.plot(x, y, style="s0.2c", pen="1p")

# raw datetime strings
x = ["2016-02-01", "2017-03-04T00:00"]
y = [7, 8]
fig.plot(x, y, style="a0.2c", pen="1p")

# the Python built-in datetime and date
x = [datetime.date(2018, 1, 1), datetime.datetime(2019, 1, 1)]
y = [8.5, 9.5]
fig.plot(x, y, style="i0.2c", pen="1p")

fig.savefig("datetime.pdf")
```
**Output:**

test_plot_datetime

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

### CODE REVIEW

**Comment by seisman** [comment]
@GenericMappingTools/python @GenericMappingTools/python-contributors This PR is almost done. If you have the GMT latest master branch installed, please try this branch and leave your comments.

**Comment by weiji14** [comment]
Some instructions on how to test this branch:

```bash
curl https://raw.githubusercontent.com/GenericMappingTools/gmt/master/ci/build-gmt-master.sh | bash
conda activate --name pygmt-test  # or some virtual env you have
pip install https://github.com/GenericMappingTools/pygmt/archive/datetime_input.zip
```
and then in python:
```
import os

os.environ["GMT_LIBRARY_PATH"] = "/home/username/gmt-install-dir/lib"

import pygmt

pygmt.print_clib_info()  # check that you're using GMT from master
```
and then try to test out `plot` with datetimes! I'll have a go at this in a bit with some 'real' data.

**Comment by weiji14** [comment]
Can't seem to get `pygmt` to wok on `gmt` master, even tried using `gmt` from bash but I'm getting an error, not sure if it's related to https://github.com/GenericMappingTools/gmt/pull/3207?

```bash
$ ~/gmt-install-dir/bin/gmt plot @DSDP.txt -R100/160/-20/30 -Jm0.3c -Sc0.2c -Gred -Bafg -pdf map
/home/username/gmt-install-dir/bin/gmt: symbol lookup error: /home/username/gmt-install-dir/bin/gmt: undefined symbol: gmt_write_glue_function
```

I'll try again later if I have time.

**Comment by seisman** [comment]
Is it possible that your `~/gmt-install-dir/bin/gmt` is loading the GMT 6.0.0 library? You can try:

1. Run `gmt --version` and check you GMT version
2. Add `-Vd` to your command, and check if there is anything that makes no sense.

**Comment by weiji14** [comment]
Ok, I tried installing it in a different conda environment and it works now! Below is my convoluted real-data example from Antarctica (longitude=46.522298201524514, latitude=-73.05774246536473).

height_over_time

Code:
<details>

```python
# %%
import os

os.environ["GMT_LIBRARY_PATH"] = "/home/username/gmt-install-dir/lib"

import numpy as np
import pygmt

pygmt.show_versions()

# %%
x = np.array(
    [
        np.datetime64(dt)
        for dt in [
            "NaT",
            "2019-01-28T07:51:16.582496785",
            "2019-04-29T03:30:57.620034382",
            "2019-07-28T23:10:36.096639410",
            "2019-10-27T18:50:32.615134843",
            "2020-01-26T14:30:18.835180975",
        ]
    ]
)
y = np.array(
    [np.NaN, 3126.60298909, 3126.67885045, 3126.67276984, 3126.7217197, 3126.68279204,]
)

# %%
# Mask out NaT values
mask = ~np.isnan(y)
x = x[mask]
y = y[mask]

# %%
dates = np.datetime_as_string(x, unit="ns")
region = [dates[0], dates[-1], np.nanmin(y) - 0.1, np.nanmax(y) + 0.1]

# %%
fig = pygmt.Figure()
fig.basemap(
    projection="X15c/5c",
    region=region,
    frame=["WSne", "xaf+lDateTime", "yaf+lHeight(m)"],
)
fig.plot(x=x, y=y, style="t1c", pen="1p")
fig.savefig("height_over_time.png")
fig.show()
```

</details>

Glad to see that these work off the shelf:
- Passing in `xarray.DataArray`s to `plot` works (thanks to [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html)'s `__array__` functionality).
- ISO datetimes down to nanoseconds are supported by GMT!

Places to improve:
- Having even just one `NaT` value will raise an error, as mentioned in https://github.com/GenericMappingTools/pygmt/issues/242#issuecomment-636269779, should be a warning ideally, but it's an upstream issue. Can be tested by commenting out the 'mask' code above.
- Support [`timedelta64`](https://pandas.pydata.org/pandas-docs/stable/user_guide/timedeltas.html), e.g. GPS time from a certain epoch. More of a nice-to-have really, could be done in a separate PR once this `datetime` one is merged in.

**Comment by seisman** [comment]
> * Passing in `xarray.DataArray`s to `plot` works (thanks to [NEP18](https://numpy.org/neps/nep-0018-array-function-protocol.html)'s `__array__` functionality).

I'll add it to the test

> * ISO datetimes down to nanoseconds are supported by GMT!

Perhaps not true. GMT converts ISO datetime to double values internally. `2019-01-28 07:51:16.582496785` is stored as a double value `1548661876.582496881484985` (printed with `%30.15lf`). So GMT only supports up to microsecond, as limited by the precision of double type. Maybe I'm wrong here.

> * The `region` argument doesn't work with numpy `datetime64` objects, have to convert to a string using [`np.datetime_as_string`](https://numpy.org/devdocs/reference/generated/numpy.datetime_as_string.html) in order to set the map frame bounds.

The decorator `kwargs_to_strings` convert the list `[w, e, s, n]` to a string *w/e/s/n*. We need to convert datetime64 object to string internally, but it's not ideal to do the conversion in the decorator. I don't have a good solution now.

> * Having even just one `NaT` value will raise an error, as mentioned in [#242 (comment)](https://github.com/GenericMappingTools/pygmt/issues/242#issuecomment-636269779), should be a warning ideally, but it's an upstream issue. Can be tested by commenting out the 'mask' code above.

Just opened an issue (https://github.com/GenericMappingTools/gmt/issues/3414) in the upstream repository.

> * Support [`timedelta64`](https://pandas.pydata.org/pandas-docs/stable/user_guide/timedeltas.html), e.g. GPS time from a certain epoch. More of a nice-to-have really, could be done in a separate PR once this `datetime` one is merged in.

Perhaps we need to convert timedelta64 to double values before passing to GMT. Leave it for a separate PR.

**Comment by weiji14** [comment]
> > * ISO datetimes down to nanoseconds are supported by GMT!
> 
> Perhaps not true. GMT converts ISO datetime to double values internally. `2019-01-28 07:51:16.582496785` is stored as a double value `1548661876.582496881484985` (printed with `%30.15lf`). So GMT only supports up to microsecond, as limited by the precision of double type. Maybe I'm wrong here.

Ideally there would be a warning, but I'm not sure if it matters. The data will be plotted in pretty much the same position anyway (unless someone looks at changes every microsecond?) so it's low priority.

> Just opened an issue ([GenericMappingTools/gmt#3414](https://github.com/GenericMappingTools/gmt/issues/3414)) in the upstream repository.

Ok, I've tested the 'new' GMT master with GenericMappingTools/gmt#3415 merged in, and it's giving a warning (as expected), but the plot shows up now :raised_hands: 

height over time, handling NaTs

Code (without mask!):

<details>

```python
import os

os.environ["GMT_LIBRARY_PATH"] = os.path.join(os.environ["HOME"], "gmt-install-dir/lib")

import numpy as np
import pygmt

pygmt.show_versions()

# %%
x = np.array(
    [
        np.datetime64(dt)
        for dt in [
            "NaT",
            "2019-01-28T07:51:16.582496785",
            "2019-04-29T03:30:57.620034382",
            "2019-07-28T23:10:36.096639410",
            "2019-10-27T18:50:32.615134843",
            "2020-01-26T14:30:18.835180975",
        ]
    ]
)
y = np.array(
    [np.NaN, 3126.60298909, 3126.67885045, 3126.67276984, 3126.7217197, 3126.68279204,]
)

# %%
dates = np.datetime_as_string(x, unit="ns")
region = [dates[1], dates[-1], np.nanmin(y) - 0.1, np.nanmax(y) + 0.1]

# %%
fig = pygmt.Figure()
fig.basemap(
    projection="X15c/5c",
    region=region,
    frame=["WSne", "xaf+lDateTime", "yaf+lHeight(m)"],
)
fig.plot(x=x, y=y, style="t1c", pen="1p")
fig.savefig("height_over_time.png")
fig.show()
```

</details>

**Comment by seisman** [comment]
> Passing in xarray.DataArrays to plot works (thanks to NEP18's __array__ functionality).

@weiji14 Could you help provide a simple example for this?

**Comment by weiji14** [comment]
> > Passing in xarray.DataArrays to plot works (thanks to NEP18's **array** functionality).
> 
> @weiji14 Could you help provide a simple example for this?

Sure, it's actually easy to construct one by wrapping a pandas.Series or numpy.array. I was trying to get a working example but the recent change you made seem to broke it? 

```python
x = xr.DataArray(data=pd.date_range(start="2020-01", periods=3))
array_to_datetime(x)
```
errors with:

```python-traceback
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-23-9a26d1f39b0d> in <module>
----> 1 array_to_datetime(x)

<ipython-input-3-d6eb22a5a706> in array_to_datetime(array)
     59                 dtype='datetime64[ns]', freq=None)
     60     """
---> 61     return pd.to_datetime(array)

~/miniconda3/envs/pygmt/lib/python3.7/site-packages/pandas/core/tools/datetimes.py in to_datetime(arg, errors, dayfirst, yearfirst, utc, format, exact, unit, infer_datetime_format, origin, cache)
    752             result = _convert_and_box_cache(arg, cache_array)
    753         else:
--> 754             result = convert_listlike(arg, format)
    755     else:
    756         result = convert_listlike(np.array([arg]), format)[0]

~/miniconda3/envs/pygmt/lib/python3.7/site-packages/pandas/core/tools/datetimes.py in _convert_listlike_datetimes(arg, format, name, tz, unit, errors, infer_datetime_format, dayfirst, yearfirst, exact)
    306         if not isinstance(arg, (DatetimeArray, DatetimeIndex)):
    307             try:
--> 308                 return DatetimeIndex(arg, tz=tz, name=name)
    309             except ValueError:
    310                 pass

~/miniconda3/envs/pygmt/lib/python3.7/site-packages/pandas/core/indexes/datetimes.py in __new__(cls, data, freq, tz, normalize, closed, ambiguous, dayfirst, yearfirst, dtype, copy, name)
    251             dayfirst=dayfirst,
    252             yearfirst=yearfirst,
--> 253             ambiguous=ambiguous,
    254         )
    255 

~/miniconda3/envs/pygmt/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py in _from_sequence(cls, data, dtype, copy, tz, freq, dayfirst, yearfirst, ambiguous)
    318             dayfirst=dayfirst,
    319             yearfirst=yearfirst,
--> 320             ambiguous=ambiguous,
    321         )
    322 

~/miniconda3/envs/pygmt/lib/python3.7/site-packages/pandas/core/arrays/datetimes.py in sequence_to_dt64ns(data, dtype, copy, tz, dayfirst, yearfirst, ambiguous)
   1795         result = result.copy()
   1796 
-> 1797     assert isinstance(result, np.ndarray), type(result)
   1798     assert result.dtype == "M8[ns]", result.dtype
   1799 

AssertionError: <class 'xarray.core.dataarray.DataArray'>
```

> And what's the best place to add some examples for calendar plots?

Putting it in the gallery should be fine in this PR. It would be nice though if we have a full tutorial for the secondary axis (make it into a `help wanted` issue)? Could repurpose the example I made for `pygmt.config` at https://github.com/GenericMappingTools/pygmt/blob/0267dd15c7626f5903cdb5a3e12ca545d70c884a/pygmt/tests/test_config.py#L67-L81.

Format time map example

**Comment by weiji14** [comment]
Thanks for pointing that out, I've added an `xarray.DataArray` example to 'test_plot_datetime' now (and updated the top post). I can add a 'test_plot3d_datetime' too after #471. Just need to wait for GMT 6.1.0 (and various other things).

**Comment by leouieda** [comment]
This is awesome @seisman! Thanks for doing it.

File: pygmt/clib/session.py [hunk 1]
```diff
@@ -48,12 +49,13 @@
...
-    "int32": "GMT_INT",
-    "uint64": "GMT_ULONG",
-    "uint32": "GMT_UINT",
+    np.float64: "GMT_DOUBLE",
+    np.float32: "GMT_FLOAT",
+    np.int64: "GMT_LONG",
+    np.int32: "GMT_INT",
+    np.uint64: "GMT_ULONG",
+    np.uint32: "GMT_UINT",
+    np.datetime64: "GMT_DATETIME",
```
Comment (weiji14) [inline pygmt/clib/session.py:58]: Quick question, is there a GMT dtype for string types? I can't seem to find one at https://docs.generic-mapping-tools.org/6.1/api.html#gmt-c-api.

Comment (seisman) [reply pygmt/clib/session.py:58]: There is a `GMT_TEXT`, but it may not related to the GMT_Put_Strings function.

**Comment by seisman** [comment]
> I'll help to fix the tests in a few minutes, seems to be just whitespace and a wrong error captured.

Thanks for help improving this PR. There are still something to do (e.g., add examples, pass a datetime region as a list), but we can merge this PR and open issues for the missing feature.

**Pull request merged**